src/Utility/Hash.php

```diff
@@ -1146,6 +1147,211 @@ public static function mergeDiff(array $data, array $compare): array
+    {
+        $fields = $schema['fields'] ?? [];
+        if (!is_array($fields) || empty($fields)) {
+            throw new CakeException('Schema is missing `fields`.');
+        }
+
+        return static::validateArray($data, ['fields' => $fields], '', $strict);
+    }
+
+    /**
+     * Validate array and walks nested arrays recursively.
+     *
+     * @param array $data Array data
+     * @param array $arraySpec Array spec
+     * @param string $fieldPrefix Dot notation prefix of field
+     * @param bool $strict Whether all keys in $data must exist in $arraySpec
+     * @return array
+     */
+    protected static function validateArray(array $data, array $arraySpec, string $fieldPrefix, bool $strict): array
+    {
+        $fields = $arraySpec['fields'] ?? [];
+        if (!is_array($fields) || empty($fields)) {
+            throw new CakeException(sprintf('Invalid array-spec for field `%s`.', substr($fieldPrefix, 0, -1)));
+        }
+
+        $errors = [];
+        foreach ($data as $key => $value) {
+            $errors += static::validateField($value, $key, $fields, $fieldPrefix, $strict);
+        }
+
+        $missing = array_diff(array_keys($fields), array_keys($data));
+        foreach ($missing as $field) {
+            $fieldSpec = static::getFieldSpec($field, $fields, $fieldPrefix);
+            if ($fieldSpec === null) {
+                continue;
+            }
+
+            if ($fieldSpec['required']) {
+                $errors[$fieldPrefix . $field] = [
+                    'expected' => static::getUnionType($fieldSpec['type']),
+                    'actual' => null,
+                ];
+            }
+        }
+
+        return $errors;
+    }
+
+    /**
+     * Validate field in array.
+     *
+     * @param mixed $value Value to validate
+     * @param string $field Field name
+     * @param array $fields Fields array
+     * @param string $fieldPrefix Dot notation prefix of field
+     * @param bool $strict Whether all keys in $data must exist in $fields
+     * @return array
+     */
+    protected static function validateField(
+        $value,
+        string $field,
+        array $fields,
+        string $fieldPrefix,
+        bool $strict
+    ): array {
+        $fieldSpec = static::getFieldSpec($field, $fields, $fieldPrefix);
+        if ($fieldSpec === null) {
+            if ($strict) {
+                return [$fieldPrefix . $field => ['expected' => null, 'actual' => getTypeName($value)]];
+            }
+
+            return [];
+        }
+
+        foreach ($fieldSpec['type'] as $typeKind => $type) {
+            switch ($typeKind) {
+                case 'array<>':
+                    if (is_array($value)) {
+                        return static::validatearray($value, $type, $fieldPrefix . $field . '.', $strict);
+                    }
+                    break;
+                case 'list<>':
+                    break;
```

Comment:
`list<>` not yet implemented. Will only support basic types like `string` and `bool`, but not `array<>`.